build/dart/rules.gni

```diff
@@ -5,7 +5,178 @@
 # This file has rules for making Dart packages and Dart-based Mojo applications.
 # The entrypoint is the dart_pkg rule.
 
+import("//build/compiled_action.gni")
 import("//build/module_args/dart.gni")
+import("//flutter/common/config.gni")
+import("//third_party/dart/build/dart/dart_action.gni")
+
+# Creates a dart kernel (dill) file suitalbe for use with gen_snapshot.
+#
+# Invoker must supply dart_main and package_config. Invoker may optionally
+# supply aot as a boolean and product as a boolean.
+template("dart_snapshot") {
```

Comment:
This is an unfortunate repetition of the rule (in [`//flutter/testing/testing.gni`](https://github.com/flutter/engine/blob/5f445d7c081116a044267c8c647bd298b06d0d13/testing/testing.gni#L192)) used by the gtest harness for Dart fixtures. Can we dry this up (in a later patch if you wish)? If anything, `//build/dart/rules.gni` is the right spot for such rules.

Reply:
I think this will lead to template name collision if both `gni`s are included from a `.GN` definition.

Reply:
There are now several places where we have rules like this - I'd like to clean them up separately to avoid potential issues that only get caught in post submit :\

One thing I can see is that the testing rules lack special handling for iOS, which will need to be addressed.